File: 2141 (Maximum Running Time of N Computers)/javascript/Solution.js
/**
 * @param {number} n
 * @param {number[]} batteries
 * @return {number}
 */
var maxRunTime = function(n, batteries) {
  let low = BigInt(1);
  let high = BigInt(0);
  for(const val of batteries){
      high+= BigInt(val);
  }
  high = high/BigInt(n);

  const possible = (mid) => {
      let totaltime= BigInt(0);
      for(const val of batteries){
          totaltime += BigInt(Math.min(val, Number(mid)));
      }
      let computer = totaltime/BigInt(mid);
      return computer>=n;
  }

  while (low<=high){
      let mid = (low+high)/BigInt(2);

      if (possible(mid)) low = mid+BigInt(1);
      else high = mid-BigInt(1);
  }
  return Number(high);
};

File: 2141 (Maximum Running Time of N Computers)/cpp/Solution.cpp
class Solution {
public:
    bool possible(long long time, int n , vector<int>& batteries){
        long long totaltime = 0;
        for(int val: batteries){
            totaltime+=min(static_cast<long long>(val), time);
        }
        long long computers = totaltime/time;
        return computers>=n;
    }
    long long maxRunTime(int n, vector<int>& batteries) {
        long long low = 1;
        long long high = 0;
        for(int val: batteries){
            high += val;
        }
        high = high/n;

        while(low<=high){
            long long mid = (low+high)>>1;
            if (possible(mid,n,batteries)) low = mid+1;
            else high = mid-1;
        }
        return high;
    }
};